File: src/main/java/me/miki/shindo/ui/comp/impl/CompDropdown.java
package me.miki.shindo.ui.comp.impl;

import lombok.Getter;
import lombok.Setter;
import me.miki.shindo.management.color.AccentColor;
import me.miki.shindo.management.color.palette.ColorPalette;
import me.miki.shindo.management.color.palette.ColorType;
import me.miki.shindo.management.nanovg.NanoVGManager;
import me.miki.shindo.management.nanovg.font.Fonts;
import me.miki.shindo.management.nanovg.font.LegacyIcon;
import me.miki.shindo.management.settings.impl.ComboSetting;
import me.miki.shindo.management.settings.impl.combo.Option;
import me.miki.shindo.ui.comp.Comp;
import me.miki.shindo.ui.framework.UIContext;
import me.miki.shindo.utils.ColorUtils;
import me.miki.shindo.utils.mouse.MouseUtils;

import java.awt.*;
import java.util.List;

public class CompDropdown extends Comp {

    private static final float CONTROL_HEIGHT = 20F;
    private static final float OPTION_HEIGHT = 18F;
    private static final float LIST_PADDING = 4F;

    private final ComboSetting setting;
    @Setter
    @Getter
    private float width;
    private boolean open;
    @Setter
    private boolean openUp;

    public CompDropdown(float x, float y, float width, ComboSetting setting) {
        super(x, y);
        this.setting = setting;
        this.width = width;
        super.setWidth(width);
        super.setHeight(CONTROL_HEIGHT);
    }

    public CompDropdown(float width, ComboSetting setting) {
        this(0F, 0F, width, setting);
    }

    public float getControlHeight() {
        return CONTROL_HEIGHT;
    }

    public boolean isOpen() {
        return open;
    }

    public void setOpen(boolean open) {
        this.open = open;
    }

    public float getDropdownHeight() {
        return open ? LIST_PADDING * 2F + Math.max(0, getOptionCount()) * OPTION_HEIGHT : 0F;
    }

    private int getOptionCount() {
        return setting != null && setting.getOptions() != null ? setting.getOptions().size() : 0;
    }

    @Override
    public void draw(int mouseX, int mouseY, float partialTicks) {
        UIContext ctx = ctx();
        NanoVGManager nvg = ctx.nvg();
        AccentColor accent = ctx.accent();
        ColorPalette palette = ctx.palette();

        float controlHeight = CONTROL_HEIGHT;
        float dropdownHeight = getDropdownHeight();

        super.setWidth(width);
        super.setHeight(controlHeight + dropdownHeight);

        float x = getX();
        float y = getY();

        nvg.drawGradientRoundedRect(x, y, width, controlHeight, 5F, accent.getColor1(), accent.getColor2());

        String label = setting != null && setting.getOption() != null ? setting.getOption().getName() : "None";
        nvg.drawText(label, x + 8F, y + 6F, Color.WHITE, 8.5F, Fonts.MEDIUM);

        String arrow = open ? LegacyIcon.CHEVRON_UP : LegacyIcon.CHEVRON_DOWN;
        nvg.drawText(arrow, x + width - 16F, y + 4F, Color.WHITE, 10F, Fonts.LEGACYICON);

        if (open && getOptionCount() > 0) {
            float listX = x;
            float listHeight = dropdownHeight;
            float listY = openUp ? y - listHeight - 4F : y + controlHeight + 4F;

            nvg.drawRoundedRect(listX, listY, width, listHeight, 5F,
                    ColorUtils.applyAlpha(palette.getBackgroundColor(ColorType.DARK), 240));

            List<Option> options = setting.getOptions();
            for (int i = 0; i < options.size(); i++) {
                Option option = options.get(i);
                float optionY = listY + LIST_PADDING + i * OPTION_HEIGHT;
                float optionHeight = OPTION_HEIGHT - 2F;
                boolean hovered = MouseUtils.isInside(mouseX, mouseY, listX + 2F, optionY, width - 4F, optionHeight);
                if (hovered) {
                    nvg.drawRoundedRect(listX + 2F, optionY, width - 4F, optionHeight, 4F,
                            ColorUtils.applyAlpha(palette.getBackgroundColor(ColorType.NORMAL), 220));
                }

                Color textColor = option.equals(setting.getOption())
                        ? palette.getFontColor(ColorType.DARK)
                        : palette.getFontColor(ColorType.NORMAL);
                nvg.drawText(option.getName(), listX + 8F, optionY + 5F, textColor, 8F, Fonts.REGULAR);
            }
        }

        super.draw(mouseX, mouseY, partialTicks);
    }

    @Override
    public void mouseClicked(int mouseX, int mouseY, int mouseButton) {
        if (mouseButton != 0) {
            return;
        }

        float controlHeight = CONTROL_HEIGHT;
        float listX = getX();
        float listHeight = getDropdownHeight();
        float listY = openUp ? getY() - listHeight - 4F : getY() + controlHeight + 4F;
        float listWidth = width;

        if (MouseUtils.isInside(mouseX, mouseY, getX(), getY(), width, controlHeight)) {
            open = !open;
            return;
        }

        if (open && MouseUtils.isInside(mouseX, mouseY, listX, listY, listWidth, listHeight)) {
            selectOptionAt(mouseX, mouseY);
            open = false;
            return;
        }

        open = false;
    }

    private void selectOptionAt(int mouseX, int mouseY) {
        if (setting == null) {
            return;
        }
        List<Option> options = setting.getOptions();
        if (options == null || options.isEmpty()) {
            return;
        }

        float optionX = getX() + 2F;
        float listHeight = getDropdownHeight();
        float optionY = (openUp ? getY() - listHeight - 4F : getY() + CONTROL_HEIGHT + 4F) + LIST_PADDING;

        for (Option option : options) {
            if (MouseUtils.isInside(mouseX, mouseY, optionX, optionY, width - 4F, OPTION_HEIGHT - 2F)) {
                setting.setOption(option);
                break;
            }
            optionY += OPTION_HEIGHT;
        }
    }
}

File: src/main/kotlin/me/miki/shindo/ui/comp/inputs/CompDropdown.kt
package me.miki.shindo.ui.comp.inputs

import me.miki.shindo.management.color.palette.ColorType
import me.miki.shindo.management.settings.impl.ComboSetting
import me.miki.shindo.management.settings.impl.combo.Option
import me.miki.shindo.management.nanovg.font.Fonts
import me.miki.shindo.management.nanovg.font.LegacyIcon
import me.miki.shindo.ui.comp.Comp
import me.miki.shindo.utils.ColorUtils
import me.miki.shindo.utils.mouse.MouseUtils
import java.awt.Color

class CompDropdown : Comp {

    private val setting: ComboSetting
    private var width: Float
    private var open: Boolean = false
    private var openUp: Boolean = false

    constructor(x: Float, y: Float, width: Float, setting: ComboSetting) : super(x, y) {
        this.setting = setting
        this.width = width
        super.setWidth(width)
        super.setHeight(CONTROL_HEIGHT)
    }

    constructor(width: Float, setting: ComboSetting) : this(0f, 0f, width, setting)

    override fun getWidth(): Float = width

    override fun setWidth(width: Float) {
        this.width = width
        super.setWidth(width)
    }
    fun setOpenUp(openUp: Boolean) {
        this.openUp = openUp
    }

    val controlHeight: Float
        get() = CONTROL_HEIGHT

    fun isOpen(): Boolean = open

    fun setOpen(open: Boolean) {
        this.open = open
    }

    val dropdownHeight: Float
        get() = if (open) LIST_PADDING * 2f + getOptionCount().coerceAtLeast(0) * OPTION_HEIGHT else 0f

    private fun getOptionCount(): Int = setting.getOptions().size ?: 0

    override fun draw(mouseX: Int, mouseY: Int, partialTicks: Float) {
        val nvgInstance = nvg
        val accent = accent
        val paletteColors = palette

        val controlHeight = CONTROL_HEIGHT
        val dropdownHeight = dropdownHeight

        super.setWidth(width)
        super.setHeight(controlHeight + dropdownHeight)

        val x = getX()
        val y = getY()

        nvgInstance.drawGradientRoundedRect(x, y, width, controlHeight, 5f, accent.color1, accent.color2)

        val label = setting.getOption()!!.name ?: "None"
        nvgInstance.drawText(label, x + 8f, y + 6f, Color.WHITE, 8.5f, Fonts.MEDIUM)

        val arrow = if (open) LegacyIcon.CHEVRON_UP else LegacyIcon.CHEVRON_DOWN
        nvgInstance.drawText(arrow, x + width - 16f, y + 4f, Color.WHITE, 10f, Fonts.LEGACYICON)

        if (open && getOptionCount() > 0) {
            val listX = x
            val listHeight = dropdownHeight
            val listY = if (openUp) y - listHeight - 4f else y + controlHeight + 4f

            nvgInstance.drawRoundedRect(
                listX,
                listY,
                width,
                listHeight,
                5f,
                ColorUtils.applyAlpha(paletteColors.getBackgroundColor(ColorType.DARK), 240)
            )

            val options: List<Option> = setting.getOptions()
            for (i in options.indices) {
                val option = options[i]
                val optionY = listY + LIST_PADDING + i * OPTION_HEIGHT
                val optionHeight = OPTION_HEIGHT - 2f
                val hovered = MouseUtils.isInside(mouseX, mouseY, listX + 2f, optionY, width - 4f, optionHeight)
                if (hovered) {
                    nvgInstance.drawRoundedRect(
                        listX + 2f,
                        optionY,
                        width - 4f,
                        optionHeight,
                        4f,
                        ColorUtils.applyAlpha(paletteColors.getBackgroundColor(ColorType.NORMAL), 220)
                    )
                }

                val textColor = if (option == setting.getOption()) {
                    paletteColors.getFontColor(ColorType.DARK)
                } else {
                    paletteColors.getFontColor(ColorType.NORMAL)
                }
                nvgInstance.drawText(option.name, listX + 8f, optionY + 5f, textColor, 8f, Fonts.REGULAR)
            }
        }

        super.draw(mouseX, mouseY, partialTicks)
    }

    override fun mouseClicked(mouseX: Int, mouseY: Int, mouseButton: Int) {
        if (mouseButton != 0) return

        val controlHeight = CONTROL_HEIGHT
        val listX = getX()
        val listHeight = dropdownHeight
        val listY = if (openUp) getY() - listHeight - 4f else getY() + controlHeight + 4f
        val listWidth = width

        if (MouseUtils.isInside(mouseX, mouseY, getX(), getY(), width, controlHeight)) {
            open = !open
            return
        }

        if (open && MouseUtils.isInside(mouseX, mouseY, listX, listY, listWidth, listHeight)) {
            selectOptionAt(mouseX, mouseY)
            open = false
            return
        }

        open = false
    }

    private fun selectOptionAt(mouseX: Int, mouseY: Int) {
        val options: List<Option> = setting.getOptions()
        if (options.isEmpty()) return

        val optionX = getX() + 2f
        val listHeight = dropdownHeight
        var optionY = (if (openUp) getY() - listHeight - 4f else getY() + CONTROL_HEIGHT + 4f) + LIST_PADDING

        for (option in options) {
            if (MouseUtils.isInside(mouseX, mouseY, optionX, optionY, width - 4f, OPTION_HEIGHT - 2f)) {
                setting.setOption(option)
                break
            }
            optionY += OPTION_HEIGHT
        }
    }

    companion object {
        private const val CONTROL_HEIGHT = 20F
        private const val OPTION_HEIGHT = 18F
        private const val LIST_PADDING = 4F
    }
}
